android-core/src/main/java/com/mparticle/internal/ConfigManager.java

```diff
@@ -212,25 +224,92 @@ static void deleteConfigManager(Context context) {
         }
     }
 
-    void saveConfigJson(JSONObject json) {
+    /**
+     * detrmine if the stored config age is greater than mMaxConfigAge and clear it from storage if it is
+     */
+    void checkConfigStaleness() {
+        Long storageDate = getConfigTimestamp();
+        if (storageDate == null) {
+            // migration step: if the current config does not have a timestamp, set one to the current time
+            setConfigTimestamp(System.currentTimeMillis());
+            storageDate = getConfigTimestamp();
+        }
+        if (mMaxConfigAge == null || mMaxConfigAge < 0) {
+            return;
+        }
+        if (mMaxConfigAge == 0) {
+            clearConfig();
+        } else {
+            if (System.currentTimeMillis() >= storageDate + TimeUnit.SECONDS.toMillis(mMaxConfigAge)) {
+                clearConfig();
+            }
+        }
+    }
+
+    String getConfig() {
+        return sPreferences.getString(CONFIG_JSON, null);
+    }
+
+    void setConfigTimestamp(Long timestamp) {
+        sPreferences.edit()
+                .putLong(CONFIG_JSON_TIMESTAMP, timestamp)
+                .apply();
+    }
+
+    @Nullable
+    Long getConfigTimestamp() {
+        if (sPreferences.contains(CONFIG_JSON_TIMESTAMP)) {
+            return sPreferences.getLong(CONFIG_JSON_TIMESTAMP, 0);
+        } else {
+            return null;
+        }
+    }
+
+    void saveConfigJson(JSONObject json, String etag, String lastModified) throws JSONException {
```

Comment:
I like moving the responsibility to persist these here. That said - there is a slight behavior change in that we used to check if they were null. This might be better - whereby if they are null we will pass null into putString below - but then we will also get null back out whereas behavior maybe we wouldn't have? Is there any concern when these are null?

Reply:
thats a good call-out, it is a change in existing behavior. When I came across the previous logic, it looked like a bug to me, although not a particularly bad one. If we fetch "config1" which returns with "etag1", then the next time around we fetch "config2" which returns _without_ an etag, on the next request we will be using the stale "etag1" in our request header which may lead the server to believe our config is stale-er than it actually is. Net-net, the way I read it, the previous behavior may lead to us fetching more configs than we need too (the extra should have been 304s).

That said, I don't see any concern over this change. Sans any comments or notes on some unexpected behavior we were accounting for when we originally implemented this, this new version seems more "correct"